File: Trees/maximum_depth_of_binary_tree.py
"""
https://leetcode.com/problems/maximum-depth-of-binary-tree/description/
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution: #Time Complexity O(n), Space Complexity O(n)
    def maxDepth(self, root: Optional[TreeNode]) -> int:

        if not root:
            return 0
        
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

File: Trees/maximum_depth_of_binary_tree.js
/*
https://leetcode.com/problems/maximum-depth-of-binary-tree/description/
*/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     * Time complexity O(n), Space complexity O(n)
     */
    maxDepth(root) {
        if(root === null){
            return 0
        }

        return 1 + Math.max(this.maxDepth(root.left), this.maxDepth(root.right))
    }
}
